net: lib: nrf_cloud: update FOTA documentation (#9047)
nrfconnect/sdk-nrf: doc/nrf/libraries/networking/nrf_cloud.rst

:c:func:`nrf_cloud_service_info_json_encode` can be used to generate the proper JSON data to enable FOTA.
Additionally, :c:func:`nrf_cloud_shadow_device_status_update` can be used to generate the JSON data and perform the shadow update.

Following are the supported FOTA types:

* ``"APP"`` - updates the application.

> * ``"BOOT"`` - updates the `upgradable second-stage bootloader`_.

Reviewer: ```suggestion
* ``"BOOT"`` - updates the :ref:`upgradable bootloader`.
```
We have a tag in the target document page, you can use it and dump the entry in the links.txt file.
(Sorry for not spotting this earlier.)

Author: This is not working:

```
Warning, treated as error:
/home/runner/work/sdk-nrf/sdk-nrf/ncs/nrf/doc/_build/nrf/src/libraries/networking/nrf_cloud.rst:139:undefined label: 'upgradable bootloader'
```

Reviewer: It needs an underscore between the words -> upgradable_bootloader.

:c:func:`nrf_cloud_service_info_json_encode` can be used to generate the proper JSON data to enable FOTA.
Additionally, :c:func:`nrf_cloud_shadow_device_status_update` can be used to generate the JSON data and perform the shadow update.

Following are the supported FOTA types:

> 

Reviewer 2: Rather than just straight up list the FOTA type strings, I like the sentences you wrote regarding FMFU and DMFU.

My suggestion, here, would be to change:

```
* ``APP``
* ``MODEM``
* ``BOOT``
* ``MDM_FULL``
```

to:
```
* :ref:`Full modem FOTA updates <full_dfu>` (enabled by ``"MDM_FULL"``) update the entire :ref:`modem firmware <architecture>` image all at once, and require |external_flash_size| of available space.
* :ref:`Delta modem FOTA updates <nrf_modem_delta_dfu>`, (enabled by ``"MODEM"``) apply incremental updates (deltas) between specific versions of  :ref:`modem firmware <architecture>`, and are much smaller in size and do not require external memory.
* :ref:`Bootloader FOTA updates <lib_dfu_target_mcuboot_style_update>` (enabled by ``"BOOT"``) update the :ref:`Upgradable second-stage bootloaders <upgradable_bootloader>`.
* `Application FOTA updates <lib_dfu_target_mcuboot_style_update>` (enabled by ``"APP"``) update the application image.
```

Note that for some reason the `MCUboot-style upgrades` section of `dfu_target.rst` (which appears to be the only page describing app & bootloader FOTA generically) for some reason doesn't have a hyperlink, so you'll need to add `.. _lib_dfu_target_mcuboot_style_update` there.

It would be nice, down the line, if we had some dedicated pages explaining:

- What the modem firmware is
- What Bootloader and Application FOTA upgrades (separately) are

I know that most of this information can be pieced together by cross-referencing a lot of different documents, but it's very spread-thin, and each fragment of information is pretty isolated. I think we would do well to try to speak about these topics holistically in one or two places, even if it is technically a bit redundant and this seems like an ideal place to do so.

Reviewer 2: If you follow this suggestion, you could ignore my suggestions about the size requirements later in this file, and actually just straight-up delete those lines

Reviewer 2: Double check that `<full_dfu>`, `<architecture>`, `<upgradable_bootloader>` actually point to `nrfxlib/nrf_modem/doc/full_dfu.rst`,`nrfxlib/nrf_modem/doc/architecture.rst`,`nrf/doc/nrf/ug_bootloader.rst`. Those three seem suspiciously non-descript, I'm not sure they'll actually work

Author: ive implemented a modified version of your suggestions

               ]
   }}}}}

Full modem FOTA updates, enabled by ``"MDM_FULL"``, require |external_flash_size| of available space.

> For the nRF9160, a full modem firmware image is approximately 2MB. Consider the power and network costs before deploying full modem FOTA updates.

Reviewer: ```suggestion
For the nRF9160, a full modem firmware image is approximately 2 MB.
Consider the power and network costs before deploying full modem FOTA updates.
```

Delta modem FOTA updates, enabled by ``"MODEM"``, are much smaller in size and do not require external memory.

You can initiate FOTA updates through `nRF Cloud`_ or by using the `nRF Cloud Device API`_.
When the device receives FOTA update information from nRF Cloud, the :c:enumerator:`NRF_CLOUD_EVT_FOTA_START` event is sent to the application.

> The FOTA update is in progress until the application receives either the :c:enumerator:`NRF_CLOUD_EVT_FOTA_DONE` event or the :c:enumerator:`NRF_CLOUD_EVT_FOTA_ERROR` event.
> Upon receipt of the :c:enumerator:`NRF_CLOUD_EVT_FOTA_DONE` event, the application must perform any necessary cleanup and perform a reboot to complete the update.

Reviewer: ```suggestion
The FOTA update is in progress until the application receives either the :c:enumerator:`NRF_CLOUD_EVT_FOTA_DONE` or :c:enumerator:`NRF_CLOUD_EVT_FOTA_ERROR` event.
When receiving the :c:enumerator:`NRF_CLOUD_EVT_FOTA_DONE` event, the application must perform any necessary cleanup and reboot the device to complete the update.
```

The message payload of the :c:enumerator:`NRF_CLOUD_EVT_FOTA_DONE` event contains the :c:enum:`nrf_cloud_fota_type` value.
If the value equals :c:enumerator:`NRF_CLOUD_FOTA_MODEM_DELTA`, the application can optionally avoid a reboot by reinitializing the modem library and then calling the :c:func:`nrf_cloud_modem_fota_completed` function.

> See `nRF Cloud FOTA`_ for details on how nRF Cloud's FOTA service operates.

Reviewer: ```suggestion
See `nRF Cloud FOTA`_ for details on the FOTA service in nRF Cloud.
```

   }}}}}

You can initiate FOTA updates through `nRF Cloud`_ or by using the `nRF Cloud Device API`_.

> When the device receives FOTA update information from nRF Cloud, the :c:enumerator:`NRF_CLOUD_EVT_FOTA_START` event is sent to the application.

Author: no, the events are sent from (this) nrf_cloud fw library.
the FOTA update information from nRF Cloud is the signal to start.

> When the device receives FOTA update information from nRF Cloud, the :c:enumerator:`NRF_CLOUD_EVT_FOTA_START` event is sent to the application.

Reviewer 2: How about:

```
When the device receives FOTA update information from nRF Cloud, the :c:enumerator:`NRF_CLOUD_EVT_FOTA_START` event is sent to the application, and the application must now perform a reboot to complete the update.
```

> When the device receives FOTA update information from nRF Cloud, the :c:enumerator:`NRF_CLOUD_EVT_FOTA_START` event is sent to the application.

Reviewer 3: How about what Georges wrote, minus the reboot comment, since you removed that as part of your PR anyway?

```
When the device receives FOTA update information from nRF Cloud, the nRF Cloud library sends the :c:enumerator:`NRF_CLOUD_EVT_FOTA_START` event to the application.
```
Alternatively, say the same, but add:
```
to the application's registered FOTA callback.
```

> When the device receives FOTA update information from nRF Cloud, the :c:enumerator:`NRF_CLOUD_EVT_FOTA_START` event is sent to the application.

Author: applied. 

`registered FOTA callback.` - this does not exist.  there is just one callback/event handler for all nrf_cloud lib events.

the reboot comment is still there, _after_ the DONE event is received.

Alternatively, you can use the :file:`app_update.bin` file to create an update bundle, but you need to enter the ``Name`` and ``Version`` fields manually.
See `nRF Cloud Getting Started FOTA documentation`_ to learn how to create an update bundle.

> Modem firmware is controlled by Nordic.

Reviewer: ```suggestion
Modem firmware is controlled by Nordic Semiconductor.
```
